### PR TITLE
KTOR-503 Add request validation plugin

### PR DESCRIPTION
**Pull request by rsinukov**
https://youtrack.jetbrains.com/issue/KTOR-503

 ```
 install(RequestValidation) {
     validate<String> {
         if (!it.startsWith("+")) ValidationResult.Invalid("$it should start with \"+\"")
         else ValidationResult.Valid
     }
 }
 install(StatusPages) {
     exception<RequestValidationException> { call, cause ->
         call.respond(HttpStatusCode.BadRequest, cause.reasons.joinToString())
     }
 }

### CODE REVIEW

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+/**
+ * A validator that should be registered with [RequestValidation] plugin
+ */
+public interface Validator {
+    /**
+     * Validates [value]
+     */
+    public suspend fun validate(value: Any): ValidationResult
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:34]: could we drop the typeless method from the API?

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * Check if [value] should be checked by this validator
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:37]: .

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    public fun filter(value: Any): Boolean
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:39]: same here

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ * Thrown when validation fails.
+ * @property value - invalid request body
+ * @property reasons - combined reasons of all validation failures for this request
+ */
+public class RequestValidationException(public val value: Any, public val reasons: List<String>) :
+    IllegalArgumentException(
+        "Validation failed for $value. Reasons: ${reasons.joinToString(".")}"
+    )
+
+private object RequestBodyTransformed : Hook<suspend (content: Any) -> Unit> {
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:87]: Could you tell me the difference between `onCallRespons.afterTransform`?

**Review by e5l — CHANGES_REQUESTED** [review]
LGTM, let's discuss the method with `Any`.
@andreyaksenov could you check the KDocs?

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import io.ktor.server.application.*
+import io.ktor.server.request.*
+
+/**
+ * A result of validation.
+ */
+public sealed class ValidationResult {
+    /**
+     * Represents successful result of validation
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:15]: > A successful result of validation.

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    public object Valid : ValidationResult()
+
+    /**
+     * Represents not successful result of validation
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:20]: > An unsuccessful result of validation

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        public constructor (reason: String) : this(listOf(reason))
+    }
+}
+
+/**
+ * A validator that should be registered with [RequestValidation] plugin
+ */
+public interface Validator {
+    /**
+     * Validates [value]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:32]: > Validates the [value].

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt (hunk 2 above)
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:37]: > Checks if the [value] should be checked by this validator.

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    public suspend fun validate(value: Any): ValidationResult
+
+    /**
+     * Check if [value] should be checked by this validator
+     */
+    public fun filter(value: Any): Boolean
+}
+
+/**
+ * A plugin that checks request body using [Validator]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:43]: > A plugin that checks a request body using [Validator].

**Review by andreyaksenov — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt (hunk 4 above)
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-request-validation/jvmAndNix/src/io/ktor/server/plugins/requestvalidation/RequestValidation.kt:87]: this is for request

**Review by e5l — APPROVED** [review]
lgtm

**Review by andreyaksenov — APPROVED** [review]

**Pull request merged**